### ISSUE TITLE
AttributeError: 'Jinja2' object has no attribute 'engine' (Django 1.8.5 compress command, having trouble)

### ISSUE THREAD

**Issue by nwilson5** [django-compressor/django-compressor#693]
Using django-compressor (1.5) extension with Jinja2 template loader on Django 1.8.5 - when I attempt to compress my css/js using:

python manage.py compress --engine jinja2

I get the following traceback:

```
Traceback (most recent call last):
  File "manage.py", line 10, in <module>
    execute_from_command_line(sys.argv)
  File "/home/user/workspace/myproject/.venv/lib/python3.4/site-packages/django/core/management/__init__.py", line 351, in execute_from_command_line
    utility.execute()
  File "/home/user/workspace/myproject/.venv/lib/python3.4/site-packages/django/core/management/__init__.py", line 343, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "/home/user/workspace/myproject/.venv/lib/python3.4/site-packages/django/core/management/base.py", line 394, in run_from_argv
    self.execute(*args, **cmd_options)
  File "/home/user/workspace/myproject/.venv/lib/python3.4/site-packages/django/core/management/base.py", line 445, in execute
    output = self.handle(*args, **options)
  File "/home/user/workspace/myproject/.venv/lib/python3.4/site-packages/django/core/management/base.py", line 661, in handle
    return self.handle_noargs(**options)
  File "/home/user/workspace/myproject/.venv/lib/python3.4/site-packages/compressor/management/commands/compress.py", line 283, in handle_noargs
    self.compress(sys.stdout, **options)
  File "/home/user/workspace/myproject/.venv/lib/python3.4/site-packages/compressor/management/commands/compress.py", line 134, in compress
    for loader in self.get_loaders():
  File "/home/user/workspace/myproject/.venv/lib/python3.4/site-packages/compressor/management/commands/compress.py", line 82, in get_loaders
    template_source_loaders.extend(e.engine.get_template_loaders(e.engine.loaders))
AttributeError: 'Jinja2' object has no attribute 'engine'
```

In `compressor.management.commands.compress.py` function `get_loaders`
```
    def get_loaders(self):
        if django.VERSION < (1, 8):
            from django.template.loader import template_source_loaders
            if template_source_loaders is None:
                try:
                    from django.template.loader import (
                        find_template as finder_func)
                except ImportError:
                    from django.template.loader import (
                        find_template_source as finder_func)  # noqa
                try:
                    # Force django to calculate template_source_loaders from
                    # TEMPLATE_LOADERS settings, by asking to find a dummy template
                    source, name = finder_func('test')
                except django.template.TemplateDoesNotExist:
                    pass
                # Reload template_source_loaders now that it has been calculated ;
                # it should contain the list of valid, instanciated template loaders
                # to use.
                from django.template.loader import template_source_loaders
        else:
            from django.template import engines
            template_source_loaders = []
            for e in engines.all():
                template_source_loaders.extend(e.engine.get_template_loaders(e.engine.loaders))
```

When using the Jinja2 backend in Django 1.8.5, the backend does not have any attribute `e.engine`. It does have `e.env.loader`. Unless I'm doing something wrong.

**Comment by diox** [comment]
Can you see if the code in #617 would fix your issue ? Some feedback in that pull request would be a great way to move things forward, thanks.

**Comment by nwilson5** [comment]
It did not.

Sorry, I tried a bit back but didn't have time to see why it wasn't working.

One question. What is the goal with Django Compressor regarding custom template backends?
The solution proposed, in the case for engine jinja2, the settings.COMPRESS_JINJA2_GET_ENVIRONMENT() (which defaults to compressor.conf.JINJA2_GET_ENVIRONMENT I believe).

The default JINJA2_GET_ENVIRONMENT simply imports jinja2 and returns jinja2.Environment(). The problem I'm having now is that this environment has no loader, thus env.list_templates is erroring since list_templates uses loader which is None.

I am using a template backend that subclasses django.template.backends.jinja2.Jinja2, although it doesn't do much to it. Although, using that backend you can specify an OPTIONS variable that defines a different environment, one that differs from jinja2.Environment. I suppose in that case, you'd have to define your own COMPRESS_JINJA2_GET_ENVIRONMENT function in settings to handle that.

If Django compressor is only trying to cover the two cases of built-in Django templates and built-in Jinja2 templates, then this solution almost works - but it ignores any custom settings (like environment) and I couldn't get it to work without redefining my own version of JINJA2_GET_ENVIRONMENT in settings, which I hacked and won't cover other cases of subclassed Jinja2 Backends


```
def COMPRESS_JINJA2_GET_ENVIRONMENT():
    import jinja2
    from django.template.loader import _engine_list
    engines = _engine_list('Jinja2')
    if engines:
        engine = engines[0]
        return engine.env
    return jinja2.Environment()
```
the default in compressor/conf.py
```
def JINJA2_GET_ENVIRONMENT():
    import jinja2
    return jinja2.Environment()
```
Which when that environment is used for loader, it will error.

This function, though, won't work if someone sub-classes Jinja2 (like I did) and uses a backend that varies slightly from the default `django.template.backends.jinja2.Jinja2`.

I'd say I don't like this solution because a backend that subclasses Jinja2 would not work by default.

I have a separate implementation that doesn't use COMPRESS_JINJA2_GET_ENVIRONMENT [EDIT: it doesn't until it tries to __get_parser()... which does use it! so either way I spin this, I need that function at the moment] in the compress command and instead loops through the engines and sees if the engine has environment already initialized. For example where he appended templates to the templates list:

```
        def _path_has_right_extension(_path, _extensions):
            return os.path.splitext(_path)[-1] in _extensions

        from django.template import engines
        for e in engines.all():
            env = e.env if hasattr(e, 'env') else None
            if env and hasattr(env, 'loader'):
                loader = env.loader
                if hasattr(loader, 'list_templates') and hasattr(loader, 'get_source'):
                    # e.g. jinja2.loaders.FileSystemLoader
                    templates |= {loader.get_source(env, template)[1]
                                  for template in loader.list_templates()
                                  if _path_has_right_extension(template, extensions)}
```
(essentially exactly what that commit is doing, except looping through all engines instead of just using the ONE env returned by JINJA2_GET_ENVIRONMENT)

E: I'm realizing after writing all of this... Is the compressor only intending to compress either `django` templates or `jinja2` templates, not both at the same time? and not intended for custom template backends? If so, keeping his changes and changing JINJA2_GET_ENVIRONMENT is all you'd need. Otherwise, if someone may have multiple jinja2-based template backends I don't think this solution will work because it assumes just one. I don't have a solution for that at the moment and don't have more time to mess with it.

I've some experience using django and django compressor but no deep understanding of how things are intended to work, so I apologize if anything I said wasn't clear. Just stating my observations after messing around with it for a little bit. I did get it working on my code base using a custom jinja2 backend that subclasses the built-in jinja2 backend.

**Comment by diox** [comment]
Ideally, compressor would work out of the box with django/jinja2 template backends (both at the same time and/or independently). 

Making it work for *custom* backends out of the box is out of scope, but it'd be nice it it provided the necessary entry points for developers wanting to do that without having to hack on compressor itself.

However, note that although I do maintain compressor, I didn't write the Jinja2 support, so I'm not super familiar with the current implementations and limitations.

**Comment by nwilson5** [comment]
In that case, I don't know what modifications that someone could make to the environment or using a backend that may end up causing an error with the compressor, but basic modifications to / subclassing of the Jinja2 built-in backend, should be able to be handled by updating the function COMPRESS_JINJA2_GET_ENVIRONMENT in one's settings.


Onto having it work on either django/jinja2 (or both), how is that intended to work?

`python manage.py compress`
`python manage.py compress --engine jinja2`
`python manage.py compress --engine django`

Is it intended that the first command would do both (if both exists else just one), the 2nd would only do jinja2, and the 3rd would only do django templates? As it stands with that commit, I don't believe it'll compress jinja2 templates unless the argument --engine explicitly states jinja2. Although I believe it will, as is, always compress django templates.

**Comment by diox** [comment]
You're right about the engines, yeah. I did say ideally :)

**Comment by nwilson5** [comment]
@diox 
I realize, after looking at the command further, that currently compress grabs either `parser` as either the django template engine parser or the jinja2 parser, not both. Thus, I believe, it can only compress one or the other at a time (defaulting to django) at the moment. I made a commit on my code that works for me, and I believe should work in general. It breaks up grabbing templates for django and jinja2 templates depending on what `engine` is.

If `engine` is defined as jinja2, it will only grab jinja2 templates. etc

If it's desired that compressor handles both at the same time, that'd require using a separate parser for separate lists of templates, which I don't feel like making work - I'd probably break something.

This commit should work for jinja2, and do either django or jinja2 compress but not both at the same time.
https://github.com/nwilson5/django-compressor/commit/ceed2af834bb0ad91775f2e348a526c053fc2cc7

**Comment by diox** [comment]
The principle looks good. I'll try to review this shortly, thanks.

**Comment by diox** [comment]
Actually, @nwilson5 : can you submit a pull request with that change ?

**Comment by nwilson5** [comment]
https://github.com/django-compressor/django-compressor/pull/713

**Comment by diox** [comment]
Should be fixed in #715, thanks.
